sim/midas/src/main/scala/midas/passes/fame/ExtractModel.scala

```diff
@@ -5,30 +5,36 @@ package midas.passes.fame
 import java.io.{PrintWriter, File}
 
 import firrtl._
-import ir._
-import Mappers._
-import Utils._
+import firrtl.ir._
+import firrtl.Utils._
 import firrtl.passes.MemPortUtils
+import firrtl.analyses.InstanceGraph
 import annotations.{InstanceTarget, Annotation, SingleTargetAnnotation}
 
 import midas.targetutils.FirrtlFAMEModelAnnotation
 
+import scala.annotation.tailrec
 import scala.collection.mutable
 import mutable.{LinkedHashSet, LinkedHashMap}
 
 class ExtractModel extends Transform {
   def inputForm = HighForm
   def outputForm = HighForm
 
-  def promoteModels(state: CircuitState): CircuitState = {
-    val anns = state.annotations.flatMap {
-      case a @ FirrtlFAMEModelAnnotation(it) if (it.module != it.circuit) => Seq(a, PromoteSubmoduleAnnotation(it))
-      case a => Seq(a)
-    }
-    if (anns.toSeq == state.annotations.toSeq) {
-      state
-    } else {
-      promoteModels((new PromoteSubmodule).runTransform(state.copy(annotations = anns)))
+  @tailrec
+  private def promoteModels(state: CircuitState): CircuitState = {
+    val fmAnns = state.annotations.collect({
+      case ann: FirrtlFAMEModelAnnotation => ann.target.module -> ann
+    }).toMap
+    // Pick by order of parent in linearization -- don't pick children of main
+    val modOrder = (new InstanceGraph(state.circuit)).moduleOrder.filterNot(_.name == state.circuit.main)
```

Comment:
Just to confirm, this is going to promote models closer to the root of the hierarchy first. Nested models are promoted later to support better composability.

Reply:
Yes, this prevents "down then up" IOs that enter the places the models got extracted from just to leave again.